async-nats/src/lib.rs

```diff
@@ -549,58 +565,75 @@ impl Connector {
                         None => {
-                            println!("Sender closed");
-                            // Sender dropped, return.
-                            break
+                            break;
                         }
                     }
                 }
 
-                result = self.connection.read_op().fuse() => {
-                    if let Ok(maybe_op) = result {
-                        match maybe_op {
-                            Some(ServerOp::Ping) => {
-                                self.connection.write_op(ClientOp::Pong).await?;
-                            }
-                            Some(ServerOp::Message { sid, subject, reply, payload }) => {
-                                let mut context = self.subscription_context.lock().await;
-                                if let Some(subscription) = context.get(sid) {
-                                    let message = Message {
-                                        subject,
-                                        reply,
-                                        payload,
-                                    };
-
-                                    subscription.sender.send(message).await.unwrap();
-                                }
-                            }
-
-                            None => {
-                                return Ok(())
-                            }
-
-                            _ => {
-                                // ignore.
-                            }
-                        }
+                maybe_op_result = self.connection.read_op().fuse() => {
+                    match maybe_op_result {
+                        Ok(Some(server_op)) => self.handle_server_op(server_op).await,
+                        _ => self.handle_reconnect().await,
                     }
                 }
-
-
             }
-            // ...
         }
 
         self.connection.stream.flush().await?;
 
         Ok(())
     }
+
+    async fn handle_server_op(&mut self, server_op: ServerOp) {
+        match server_op {
+            ServerOp::Ping => {
+                self.connection.write_op(ClientOp::Pong).await.unwrap();
+            }
+
+            ServerOp::Message {
+                sid,
+                subject,
+                reply,
+                payload,
+            } => {
+                let mut context = self.subscription_context.lock().await;
+                if let Some(subscription) = context.get(sid) {
+                    let message = Message {
+                        subject,
+                        reply,
+                        payload,
+                    };
+
+                    subscription.sender.send(message).await.unwrap();
+                }
+            }
+            _ => {}
+        }
+    }
+
+    async fn handle_client_op(&mut self, client_op: ClientOp) {
+        if let Err(err) = self.connection.write_op(client_op).await {
```

Comment:
CllientOp being moved currently prohibits this:
 
 ```suggestion
        while let Err(err) = self.connection.write_op(client_op).await {
```